src/3d/qgsoffscreen3dengine.cpp

```diff
@@ -0,0 +1,230 @@
+#include <Qt3DRender/QRenderCapture>
+#include <Qt3DRender/QRenderSettings>
+#include <Qt3DRender/QRenderTarget>
+#include <Qt3DRender/QRenderTargetOutput>
+#include <Qt3DRender/QRenderTargetSelector>
+#include <Qt3DRender/QRenderSurfaceSelector>
+#include <Qt3DRender/QTexture>
+#include <Qt3DRender/QViewport>
+
+
+QgsOffscreen3DEngine::QgsOffscreen3DEngine()
+{
+  mSize = QSize( 640, 480 );
+
+  // Set up the default OpenGL surface format.
+  QSurfaceFormat format;
+  format.setDepthBufferSize( 32 ); // TODO: or 24?  (used by QWindow3D)
+  format.setSamples( 8 );
+  QSurfaceFormat::setDefaultFormat( format );
+
+  // Set up a camera to point at the shapes.
+  mCamera = new Qt3DRender::QCamera;
+  mCamera->lens()->setPerspectiveProjection( 45.0f, float( mSize.width() ) / float( mSize.height() ), 0.1f, 1000.0f );
+  mCamera->setPosition( QVector3D( 0, 0, 20.0f ) );
+  mCamera->setUpVector( QVector3D( 0, 1, 0 ) );
+  mCamera->setViewCenter( QVector3D( 0, 0, 0 ) );
+
+  // Set up the engine and the aspects that we want to use.
+  mAspectEngine = new Qt3DCore::QAspectEngine();
+  mRenderAspect = new Qt3DRender::QRenderAspect( Qt3DRender::QRenderAspect::Threaded ); // Only threaded mode seems to work right now.
+  mLogicAspect = new Qt3DLogic::QLogicAspect();
+
+  mAspectEngine->registerAspect( mRenderAspect );
+  mAspectEngine->registerAspect( mLogicAspect );
+
+  // Create the root entity of the engine.
+  // This is not the same as the 3D scene root: the QRenderSettings
+  // component must be held by the root of the QEntity tree,
+  // so it is added to this one. The 3D scene is added as a subtree later,
+  // in setRootEntity().
+  mRoot = new Qt3DCore::QEntity();
+  mRenderSettings = new Qt3DRender::QRenderSettings( mRoot );
+  mRoot->addComponent( mRenderSettings );
+
+  mCamera->setParent( mRoot );
+
+  // Create the offscreen frame graph, which will manage all of the resources required
+  // for rendering without a QWindow.
+  createFrameGraph();
+
+  // Set this frame graph to be in use.
+  mRenderSettings->setActiveFrameGraph( mSurfaceSelector );
+
+  // Set the root entity of the engine. This causes the engine to begin running.
+  mAspectEngine->setRootEntity( Qt3DCore::QEntityPtr( mRoot ) );
+
+}
+
+QgsOffscreen3DEngine::~QgsOffscreen3DEngine()
+{
+  delete mAspectEngine;
+}
+
+void QgsOffscreen3DEngine::setSize( const QSize &s )
+{
+  mSize = s;
+
+  mTexture->setSize( mSize.width(), mSize.height() );
+  mDepthTexture->setSize( mSize.width(), mSize.height() );
+  mSurfaceSelector->setExternalRenderTargetSize( mSize );
+
+  mCamera->setAspectRatio( float( mSize.width() ) / float( mSize.height() ) );
+}
+
+void QgsOffscreen3DEngine::setClearColor( const QColor &color )
+{
+  mClearBuffers->setClearColor( color );
+}
+
+void QgsOffscreen3DEngine::setFrustumCullingEnabled( bool enabled )
+{
+  // TODO
+  Q_UNUSED( enabled );
+}
+
+void QgsOffscreen3DEngine::createRenderTarget()
+{
+  mTextureTarget = new Qt3DRender::QRenderTarget;
```

Comment:
I think this is leaked - I can't see it cleaned up anywhere

Reply:
Checked and this should get parented to mRenderTargetSelector (added a comment about it)